dhis2/dhis2-core
# fix: Scheduler - reliably remove jobs from schedule [DHIS2-14316]

### Summary
Main goal was to make the scheduler behave reasonable and consistent when removing previously scheduled jobs from the schedule. There are two aspects that need to come together consistent:

* the actual schedule managed in memory by an executor 
* the description and state of the jobs

The issues found were inconsistencies of the executor state and the state of the job descriptions.

The semantics chosen to make this reliable, predictable and consistent are as follows

* `SchedulingManager.stop`ing a `JobConfiguration` removes it from the schedule **but** will finish a run that already started
* `SchedulingManager.cancel`ing a `JobType` is the cooperative attempt to abort a running job of the type independent of the configuration that started it. This does **not** remove the job from the schedule but only aborts the current run.
* There can be multiple `JobConfiguration`s of the same `JobType` put on a schedule but only one of them will be able to run at the same time. If a job run is due to run while another of the same type already is in progress the attempt to run fails.

### Additional Improvements and Fixes
* fix: added CRON expression validation to the job queue create and update
* feat: added a `TEST` job which allows to test failing job scenarios
* feat: job configurations that are part of a queue can be executed ad-hoc using `executeNow`. This will start the queue from their position onward. This might come in handy when a queue fails at a certain points and the queue should be resumed from that point after the data has been adjusted manually
* fix: the `JobConfiguration.setLastExecuted` records the start time of the job run, not the finish time
* fix: the `STOPPED` status is persisted after a cancelled run (instead of `FAILED`)
* fix: `SchedulerEntry.of` reorders jobs in the queue by position

### Automatic Testing
New controller tests were added for the added job queue validation
New unit tests were added for `DefaultSchedulingManager` with job queues

### Manual Testing
Testing Job Queues

* create 2-3 jobs (ideally but not necessarily of different types)
* create a job queue with the jobs
* run the queue by waiting for the CRON trigger or run the first job in the queue manually (run now)
* check the job progress tracking for the jobs in the queue after the run. Compare the timestamps to verify they ran in the queue order

Testing Stop

* create a job that gives you some time to stop it while it is running (for example use TEST job with `waitMillis` or a slow job like analytics)
* start the run manually (or wait for the scheduled execution)
* while it is running stop the job by editing its configuration
* check the job ran did continue and finish
* redo steps so far except disable the job
* check the job ran did continue and finish
* check that job is no longer on schedule (ideally a CRON expression is used that allows to just wait a short time to verify it did indeed not run again, for example use ` 0/5 * * * *` for every 5 minutes)
